consensus/parlia/parlia.go

```diff
@@ -592,6 +753,109 @@ func (p *Parlia) verifySeal(chain consensus.ChainHeaderReader, header *types.Hea
 	return nil
 }
 
+func (p *Parlia) prepareValidators(chain consensus.ChainHeaderReader, header *types.Header) error {
+	if header.Number.Uint64()%p.config.Epoch != 0 {
+		return nil
+	}
+
+	newValidators, voteAddressMap, err := p.getCurrentValidators(header.ParentHash, new(big.Int).Sub(header.Number, big.NewInt(1)))
+	if err != nil {
+		return err
+	}
+	// sort validator by address
+	sort.Sort(validatorsAscending(newValidators))
+	if !p.chainConfig.IsBoneh(header.Number) {
+		for _, validator := range newValidators {
+			header.Extra = append(header.Extra, validator.Bytes()...)
+		}
+	} else {
+		header.Extra = append(header.Extra, byte(len(newValidators)))
+		for _, validator := range newValidators {
+			header.Extra = append(header.Extra, validator.Bytes()...)
+			header.Extra = append(header.Extra, voteAddressMap[validator].Bytes()...)
+		}
+	}
+	return nil
+}
+
+func (p *Parlia) assembleVoteAttestation(chain consensus.ChainHeaderReader, header *types.Header) error {
+	if !p.chainConfig.IsBoneh(header.Number) || header.Number.Uint64() < 2 {
+		return nil
+	}
+
+	if reflect.ValueOf(p.votePool).IsNil() {
+		return errors.New("vote pool is nil")
+	}
+
+	// Fetch direct parent's votes
+	parent := chain.GetHeaderByHash(header.ParentHash)
+	if parent == nil {
+		return errors.New("parent not found")
+	}
+	snap, err := p.snapshot(chain, parent.Number.Uint64()-1, parent.ParentHash, nil)
+	if err != nil {
+		return err
+	}
+	votes := p.votePool.FetchVoteByHash(parent.Hash())
+	if len(votes) <= len(snap.Validators)*2/3 {
+		return nil
+	}
+
+	// Prepare vote attestation
+	// Prepare vote data
+	justified := p.GetHighestJustifiedHeader(chain, parent)
+	if justified == nil {
+		return errors.New("highest justified block not found")
+	}
+	attestation := &types.VoteAttestation{
+		Data: &types.VoteData{
+			SourceNumber: justified.Number.Uint64(),
+			SourceHash:   justified.Hash(),
+			TargetNumber: parent.Number.Uint64(),
+			TargetHash:   parent.Hash(),
+		},
+	}
+	// Check vote data from votes
+	for _, vote := range votes {
+		if vote.Data.Hash() != attestation.Data.Hash() {
+			return fmt.Errorf("vote check error, expected: %v, real: %v", attestation.Data, vote)
+		}
+	}
+	// Prepare aggregated vote signature
+	voteAddrSet := make(map[types.BLSPublicKey]struct{}, len(votes))
```

Comment:
Will a bad validator on purpose drop some votes to get more rewards? For example, if it receive all votes, it can drop several votes (but still bigger than 3/4), then it can get more rewards. Am I correct?

Reply:
Good catch, you are right. But this is a common problem in the PoS network, after discussing it, we decide not to fix this because this won't affect our mechanism.

Reply:
OK, as a known issue